Mathlib/Topology/ContinuousFunction/CompactlySupported.lean

```diff
@@ -0,0 +1,700 @@
+  continuous_toFun := by
+    rw [h]
+    exact f.continuous_toFun
+  has_compact_support' := by
+    simp_rw [h]
+    exact f.has_compact_support'
+
+@[simp]
+theorem coe_copy (f : C_c(α, β)) (f' : α → β) (h : f' = f) : ⇑(f.copy f' h) = f' :=
+  rfl
+
+theorem copy_eq (f : C_c(α, β)) (f' : α → β) (h : f' = f) : f.copy f' h = f :=
+  DFunLike.ext' h
+
+theorem eq_of_empty [IsEmpty α] (f g : C_c(α, β)) : f = g :=
+  ext <| IsEmpty.elim ‹_›
+
+/-- A continuous function on a compact space has automatically compact support. -/
+@[simps]
+def ContinuousMap.liftCompactlySupported [CompactSpace α] : C(α, β) ≃ C_c(α, β) where
+  toFun f :=
+    { toFun := f
+      continuous_toFun := f.continuous
+      has_compact_support' := by
+        simp only
+        exact IsCompact.of_isClosed_subset isCompact_univ (isClosed_tsupport f)
+          (Set.subset_univ (tsupport f))
+        }
+  invFun f := f
+  left_inv f := by
+    ext
+    rfl
+  right_inv f := by
+    ext
+    rfl
+
+/-- A continuous function on a compact space has automatically compact support. This is not an
+instance to avoid type class loops. -/
+lemma compactlySupportedContinuousMapClass.ofCompact {G : Type*} [FunLike G α β]
+    [ContinuousMapClass G α β] [CompactSpace α] : CompactlySupportedContinuousMapClass G α β where
+  map_continuous := map_continuous
+  has_compact_support := by
+    intro f
+    exact IsCompact.of_isClosed_subset isCompact_univ (isClosed_tsupport f)
+      (Set.subset_univ (tsupport f))
+
+end Basics
+
+/-! ### Algebraic structure
+
+Whenever `β` has the structore of continuous addtive monoid and a compatible topological structure,
+then `C_c(α, β)` inherits a corresponding algebraic structure. The primary exception to this is that
+`C_c(α, β)` will not have a multiplicative identity.
+-/
+
+section AlgebraicStructure
+
+variable [TopologicalSpace β] (x : α)
+
+instance instZero [Zero β] : Zero C_c(α, β) where
+  zero := { toFun := (0 : C(α, β))
+            continuous_toFun := (0 : C(α, β)).2
+            has_compact_support' := by
+              rw [HasCompactSupport, tsupport]
+              simp only [ContinuousMap.coe_zero, Function.support_zero', closure_empty,
+                isCompact_empty]
+}
+
+instance instInhabited [Zero β] : Inhabited C_c(α, β) :=
+  ⟨0⟩
+
+@[simp]
+theorem coe_zero [Zero β] : ⇑(0 : C_c(α, β)) = 0 :=
+  rfl
+
+theorem zero_apply [Zero β] : (0 : C_c(α, β)) x = 0 :=
+  rfl
+
+instance instMul [MulZeroClass β] [ContinuousMul β] : Mul C_c(α, β) :=
+  ⟨fun f g => ⟨f * g, HasCompactSupport.mul_left g.2⟩⟩
+
+@[simp]
+theorem coe_mul [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : ⇑(f * g) = f * g :=
+  rfl
+
+@[simp]
+theorem mul_apply [MulZeroClass β] [ContinuousMul β] (f g : C_c(α, β)) : (f * g) x = f x * g x :=
+  rfl
+
+-- the product of `(f : C(α, β))` and `(g : C_c(α, β))` is in `C_c(α, β)`
+instance instSMulC [MulZeroClass β] [ContinuousMul β] : SMul C(α, β) C_c(α, β) :=
+  ⟨fun f g => ⟨f * g, HasCompactSupport.mul_left g.2⟩⟩
```

Comment:
Instead of registering this for `C(α, β)`, could you register this for any `ContinuousMapClass`? Even better, could you register this in the more general setting where you have a type gamma acting continuously on beta, for any `ContinuousMapClass` of functions from alpha to gamma?

Reply:
Could you elaborate what I should do here? I suppose I should define `SMul` between `F` and `G` with `ContinuousMapClass F α β` and `CompactlySupportedContinuousMapClass G α β`, but I don't know how to get an object of type `G` from, say, `f : F`, `g : G` and `fun x => (f x) * (g x)`.

Reply:
I mean something like
```
instance [Zero β] [TopologicalSpace γ] [SMulZeroClass γ β] [ContinuousSMul γ β]
    {F : Type*} [FunLike F α γ] [ContinuousMapClass F α γ] : SMul F C_c(α, β) where
  smul f g :=
    ⟨⟨fun x ↦ f x • g x, (map_continuous f).smul g.continuous⟩, g.hasCompactSupport'.smul_left⟩
```
The space on the right of the action is always `C_c(α, β)`, but the space on the left is any space of continuous functions taking values in a space gamma acting continuously on beta.

Note that for this to work you will need to change the assumptions in the section `DistribMulAction` of the file `Topology.Support`: it should be replaced with the more general
```
section SMulZeroClass

variable [TopologicalSpace α] [Zero M] [SMulZeroClass R M]
variable {f : α → R} {f' : α → M} {x : α}

theorem HasCompactSupport.smul_left (hf : HasCompactSupport f') : HasCompactSupport (f • f') := by
  rw [hasCompactSupport_iff_eventuallyEq] at hf ⊢
  exact hf.mono fun x hx => by simp_rw [Pi.smul_apply', hx, Pi.zero_apply, smul_zero]
#align has_compact_support.smul_left HasCompactSupport.smul_left

end SMulZeroClass
```

Reply:
I see, I just changed as you suggested.